src/systems/acoustic_comms/AcousticComms.cc

```diff
@@ -0,0 +1,204 @@
+
+    if (srcAddressAttachedToModel)
+    {
+      // All these messages need to be processed.
+      for (auto &msg : outbound)
+      {
+        // Is the destination address bound?
+        auto itDst = _currentRegistry.find(msg->dst_address());
+        bool dstAddressBound = itDst != _currentRegistry.end();
+
+        // Is the destination address attached to a model?
+        bool dstAddressAttachedToModel =
+          dstAddressBound && itDst->second.entity != kNullEntity;
+
+        if (!dstAddressAttachedToModel)
+          continue;
+
+        // The plugin checks the distance travelled by the signal
+        // so far. If it is more than the maxRange, it is dropped
+        // and would never reach the destination.
+        // If it has already reached the destination but not as far
+        // as maxRange, it is processed.
+        // If it has reached neither the destination nor the maxRange,
+        // it is considered in transit.
+
+        if (this->dataPtr->poseSrcAtMsgTimestamp.count(msg) == 0)
+        {
+          // This message is being processed for the first time.
+          // Record the current position of the sender and use it
+          // for distance calculations.
+          this->dataPtr->poseSrcAtMsgTimestamp[msg] =
+            worldPose(itSrc->second.entity, _ecm).Pos();
+        }
+
+        math::Vector3d poseSrc =
+          this->dataPtr->poseSrcAtMsgTimestamp[msg];
+
+        // Calculate distance between the bodies.
+        auto poseDst = worldPose(itDst->second.entity, _ecm).Pos();
+        auto distanceToTransmitter =
+          (poseSrc - poseDst).Length();
+
+        // Calculate distance covered by the message.
+        std::chrono::steady_clock::time_point currTime(_info.simTime);
+        auto timeOfTransmission = msg->mutable_header()->stamp();
+
+        auto currTimestamp =
+          std::chrono::nanoseconds(currTime.time_since_epoch().count());
+        auto packetTimestamp =
+          std::chrono::seconds(timeOfTransmission.sec()) +
+          std::chrono::nanoseconds(timeOfTransmission.nsec());
+
+        std::chrono::duration<double> deltaT =
+          currTimestamp - packetTimestamp;
+        double distanceCoveredByMessage = deltaT.count() *
+          this->dataPtr->speedOfSound;
+
+        // Check the msgs that haven't exceeded the maxRange.
+        if (distanceCoveredByMessage <= this->dataPtr->maxRange)
+        {
+          if (distanceCoveredByMessage >= distanceToTransmitter)
+          {
+            // This message needs to be processed.
+            _newRegistry[msg->dst_address()].inboundMsgs.push_back(msg);
+            this->dataPtr->poseSrcAtMsgTimestamp.erase(msg);
+          }
+          else
+          {
+            // This message is still in transit, should be kept in the
+            // outbound buffer of source and not moved to inbound of
+            // the destination.
+            newOutbound.push_back(msg);
+          }
+        }
+        else
+        {
```

Comment:
I think there's a situation that isn't covered here: If source and destination are at a distance > `maxRange` but the destination is moving closer to the sender (in the direction of the message). If I'm not wrong, you'll discard the message instantly but you're not accounting that potentially the destination might be within range while the message is traveling.

Reply:
I am actually, ``distanceToTransmitter`` is the distance between :
* Position of sender when the message was sent, (this is constant throughout the lifetime of the message)
* Current position of the destination. This is when the plugin is spinning, so always contains the current position of destination. So even if the destination has just moved towards the source, that distance should be reflected accurately.

Reply:
You're right, the key is to use `distanceCoveredByMessage` to decide when to drop the message and not `distanceToTransmitter` as I originally thought you were using.